File: 20200215Python-China/0104. Maximum Depth of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if not root:
            return 0
        deep=0
        root_list=[root]
        while root_list:
            deep+=1
            len_of_root_list=len(root_list)
            for i in range(len_of_root_list):
                cur_root=root_list.pop(0)
                if cur_root.left:
                    root_list.append(cur_root.left)
                if cur_root.right:
                    root_list.append(cur_root.right)
        return deep

File: archives/C#/0104. Maximum Depth of Binary Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int MaxDepth(TreeNode root) {
        if(root==null){
            return 0;
        }
        List<TreeNode> root_list= new List<TreeNode>(){root};
        List<TreeNode> newt_list= new List<TreeNode>(){};
        int rep=0;
        while(root_list.Count!=0){
            rep+=1;
            newt_list.Clear();
            foreach(TreeNode curroot in root_list){
                if (curroot.left!=null){
                    newt_list.Add(curroot.left);
                }
                if (curroot.right!=null){
                    newt_list.Add(curroot.right);
                }
            }
            root_list.Clear();
            root_list.AddRange(newt_list);
        }
        return rep;
    }
}
//==========================================================================================
public class Solution {
    public int MaxDepth(TreeNode root) {
        if(root==null){
            return 0;
        }
        return 1+Math.Max(MaxDepth(root.left),MaxDepth(root.right));
    }
}
//===========================================================================================
public class Solution {
    public int MaxDepth(TreeNode root) {
        if(root==null){
            return 0;
        }
        
        Queue<TreeNode> rootQueue=new Queue<TreeNode>();
        rootQueue.Enqueue(root);
        int rep=0;
        while(rootQueue.Count!=0){
            rep+=1;
            int lenQueue=rootQueue.Count;
            for(int i=0;i<lenQueue;i++){
                TreeNode node=rootQueue.Dequeue();
                if (node.left!=null){
                    rootQueue.Enqueue(node.left);
                }
                if (node.right!=null){
                    rootQueue.Enqueue(node.right);
                }
            }
        }
        return rep;
        
    }
}
